cli/2024-10-cli-udt.md

```diff
@@ -0,0 +1,232 @@
+
+### Top level goals
+<!-- At the most basic level, what are we trying to accomplish? -->
+- Provide an intuitive and seamless experience for platform engineers to create, update, and delete User-Defined-Type definitions using the CLI.
+- Ensure that the CLI experience is consistent with the other commands in Radius
+
+### Non-goals (out of scope)
+<!-- What are we explicitly not trying to accomplish? -->
+- Other interfaces and experiences to interact with user defined resource type definitions and resources are out of scope for this document
+
+## Key scenarios
+<!-- List ~3-7 high level scenarios to clarify the value and point to how we will decompose this big area into component capabilities. We may ultimately have more than one level of scenario. -->
+The key scenarios for UDTs are outlined [here](/architecture/2024-06-resource-extensibility-feature-spec.md). We will decompose these scenarios specific to the CLI experience.
+
+### Scenario 1: Deb defines and deploys the internal messaging service using the CLI
+<!-- One or two sentence summary -->
+Deb, a platform engineer who wants to define and deploy their internal messaging service Plaid as a resource type in Radius. 
+
+Before we start designing the experience, we need to understand the high level terminologies/concept that a user would interact with in the CLI experience.
+
+| Terminology | Description | Example | 
+|-------------|-------------|---------|
+|Resource provider | Refers to a group of resource types under a common namespace. Resource providers are also entities that implement the group of resource types | Mycompany.Messaging or Applications.Core |
+|Resource type | A service or technology that can be modeled as a resource | PostgreSQL or AWS S3 or Internal messaging service |
+|Resource | An instance of a resource type | mydb or mybucket |
+
+#### User Experience
+Below is an illustrative example of how Deb would define the internal messaging service Plaid as a resource type in Radius. 
+
+1. Deb creates the resource provider and the resource type with Radius
+
+    ```bash
+    rad resource-provider create -f Mycompany.Messaging.yaml
+    ```
+    `Mycompany.Messaging.yaml` is the schema definition file that contains the resource type definition for the messaging service Plaid. See the feature spec that covers the [authoring experience for the schema definition](https://github.com/radius-project/design-notes/blob/67dd03056d22bcae2ee9743ffb308c8ac7f012bd/architecture/2024-10-user-defined-types-authoring-experience-feature-spec.md)
+
+2. Deb creates the resource instance for the messaging service Plaid
+
+    ```bash
+    rad resource create Mycompany.Messaging/Plaid myplaid -f myplaid.json
```

Comment:
Does the resource type `Plaid` get created as a part of the previous resource provider create command?

Reply:
yes

Reply:
I'm a bit confused on this as well. We create the resource type as part of the previous command but then the resource instance here?

Reply:
yes `rad resource-type create` creates the resource type definition and `rad resource create` creates the resource instance.